File: src/main/java/de/esymetric/jerusalem/ownDataRepresentation/fileSystem/LatLonDir.java
package de.esymetric.jerusalem.ownDataRepresentation.fileSystem;

import java.io.File;

public class LatLonDir {

	public final static double LAT_OFFS = 90.0;
	public final static double LNG_OFFS = 180.0;

	int latInt;
	int lngInt;

	public LatLonDir(double lat, double lng) {
		latInt = (int) (lat + LAT_OFFS);
		lngInt = (int) (lng + LNG_OFFS);
	}

	public LatLonDir(double lat, double lng, int offsetBits) {
		latInt = (int) (lat + LAT_OFFS);
		lngInt = (int) (lng + LNG_OFFS);

		if ((offsetBits & 1) != 0)
			latInt++;
		if ((offsetBits & 2) != 0)
			latInt--;
		if ((offsetBits & 4) != 0)
			lngInt++;
		if ((offsetBits & 8) != 0)
			lngInt--;
	}

	public LatLonDir(short shortKey) {
		int key = shortKey;
		if (key < 0)
			key += 65536;
		latInt = key / 360;
		lngInt = key - 360 * latInt;
	}

	public int getOffsetBits(double lat, double lng) {
		LatLonDir lld2 = new LatLonDir(lat, lng);
		int deltaLat = lld2.latInt - latInt;
		int deltaLng = lld2.lngInt - lngInt;
		int bits = 0;
		if (deltaLat == 1)
			bits |= 1;
		if (deltaLat == -1)
			bits |= 2;
		if (deltaLng == 1)
			bits |= 4;
		if (deltaLng == -1)
			bits |= 8;
		return bits;
	}

	public short getShortKey() {
		return (short) (latInt * 360 + lngInt);
	}

	public boolean equals(LatLonDir lld) {
		return lld.latInt == latInt && lld.lngInt == lngInt;
	}

	public String getDir(String dataDirectoryPath) {
		String path = dataDirectoryPath + File.separatorChar + "lat_" + latInt
				+ File.separatorChar + "lng_" + lngInt;
		return path;
	}

	public String makeDir(String dataDirectoryPath, boolean createDir) {
		String path = dataDirectoryPath + File.separatorChar + "lat_" + latInt
				+ File.separatorChar + "lng_" + lngInt;
		if (createDir) {
			File d = new File(path);
			if( !d.exists() )
				d.mkdirs();
		}
		return path;
	}

	public static void deleteAllLatLonDataFiles(String dataDirectoryPath) {
		for (File f : new File(dataDirectoryPath).listFiles())
			if (f.isDirectory() && f.getName().startsWith("lat_"))
				for (File g : f.listFiles())
					if (g.isDirectory() && g.getName().startsWith("lng_"))
						for (File h : g.listFiles())
							if (h.isFile())
								h.delete();

	}

}

File: src/main/java/de/esymetric/jerusalem/ownDataRepresentation/fileSystem/LatLonDir.kt
package de.esymetric.jerusalem.ownDataRepresentation.fileSystem

import java.io.File

class LatLonDir {
    var latInt: Int
    var lngInt: Int

    constructor(lat: Double, lng: Double) {
        latInt = (lat + LAT_OFFS).toInt()
        lngInt = (lng + LNG_OFFS).toInt()
    }

    constructor(lat: Double, lng: Double, offsetBits: Int) {
        latInt = (lat + LAT_OFFS).toInt()
        lngInt = (lng + LNG_OFFS).toInt()
        if (offsetBits and 1 != 0) latInt++
        if (offsetBits and 2 != 0) latInt--
        if (offsetBits and 4 != 0) lngInt++
        if (offsetBits and 8 != 0) lngInt--
    }

    constructor(shortKey: Short) {
        var key = shortKey.toInt()
        if (key < 0) key += 65536
        latInt = key / 360
        lngInt = key - 360 * latInt
    }

    fun getOffsetBits(lat: Double, lng: Double): Int {
        val lld2 = LatLonDir(lat, lng)
        val deltaLat = lld2.latInt - latInt
        val deltaLng = lld2.lngInt - lngInt
        var bits = 0
        if (deltaLat == 1) bits = bits or 1
        if (deltaLat == -1) bits = bits or 2
        if (deltaLng == 1) bits = bits or 4
        if (deltaLng == -1) bits = bits or 8
        return bits
    }

    val shortKey: Short
        get() = (latInt * 360 + lngInt).toShort()

    fun equals(lld: LatLonDir?): Boolean {
        return lld!!.latInt == latInt && lld.lngInt == lngInt
    }

    fun getDir(dataDirectoryPath: String): String {
        return (dataDirectoryPath + File.separatorChar + "lat_" + latInt
                + File.separatorChar + "lng_" + lngInt)
    }

    fun makeDir(dataDirectoryPath: String, createDir: Boolean): String {
        val path = (dataDirectoryPath + File.separatorChar + "lat_" + latInt
                + File.separatorChar + "lng_" + lngInt)
        if (createDir) {
            val d = File(path)
            if (!d.exists()) d.mkdirs()
        }
        return path
    }

    companion object {
        const val LAT_OFFS = 90.0
        const val LNG_OFFS = 180.0

		fun deleteAllLatLonDataFiles(dataDirectoryPath: String?) {
            for (f in File(dataDirectoryPath).listFiles()) if (f.isDirectory && f.name.startsWith("lat_")) for (g in f.listFiles()) if (g.isDirectory && g.name.startsWith(
                    "lng_"
                )
            ) for (h in g.listFiles()) if (h.isFile) h.delete()
        }
    }
}